examples/hamiltonian_representation.py

```diff
@@ -0,0 +1,322 @@
+
+    raise ValueError(f'Unsupported type: {type(boolean_expr)}')
+
+
+def get_name_to_id(boolean_exprs: Sequence[Expr]) -> Dict[str, int]:
+    """Maps the variables to a unique integer.
+
+    Args:
+        boolean_expr: A Sympy expression containing symbols and Boolean operations
+
+    Return:
+        A dictionary of string (the variable name) to a unique integer.
+    """
+
+    # For run-to-run identicalness, we sort the symbol name lexicographically.
+    symbol_names = sorted(
+        {symbol.name for boolean_expr in boolean_exprs for symbol in boolean_expr.free_symbols}
+    )
+    return {symbol_name: i for i, symbol_name in enumerate(symbol_names)}
+
+
+def _gray_code_comparator(k1, k2, flip=False):
+    max_1 = k1[-1] if k1 else -1
+    max_2 = k2[-1] if k2 else -1
+    if max_1 != max_2:
+        return -1 if (max_1 < max_2) ^ flip else 1
+    if max_1 == -1:
+        return 0
+    return _gray_code_comparator(k1[0:-1], k2[0:-1], not flip)
+
+
+def build_circuit_from_hamiltonians(
+    hamiltonian_polynomial_list: List[HamiltonianPolynomial],
+    qubits: List[cirq.NamedQubit],
+    theta: float,
+    ladder_target: bool = False,
+) -> cirq.Circuit:
+    """Builds a circuit according to [1].
+
+    Args:
+        hamiltonian_polynomial_list: the list of Hamiltonians, typically built by calling
+            build_hamiltonian_from_boolean().
+        qubits: The list of qubits corresponding to the variables.
+        theta: A single float scaling the rotations.
+        ladder_target: Whether to use convention of figure 7a or 7b.
+
+    Return:
+        A dictionary of string (the variable name) to a unique integer.
+    """
+    combined = sum(hamiltonian_polynomial_list, HamiltonianPolynomial.O())
+
+    circuit = cirq.Circuit()
+
+    # Here we follow improvements of [4] cancelling out the CNOTs. The first step is to order by
+    # Gray code so that as few as possible gates are changed.
+    sorted_hs = sorted(
+        list(combined.hamiltonians.keys()), key=functools.cmp_to_key(_gray_code_comparator)
+    )
+
+    def _apply_cnots(ph: Tuple[int, ...], ch: Tuple[int, ...]):
+        # This function applies in sequence the CNOTs from ph and then ch. However, given that the
+        # h are sorted in Gray ordering and that some cancel each other, we can reduce the number
+        # of gates. See [4] for more details.
+
+        cnots: List[Tuple[int, int]] = []
+
+        if ladder_target:
+            cnots.extend((ph[i], ph[i + 1]) for i in reversed(range(len(ph) - 1)))
+            cnots.extend((ch[i], ch[i + 1]) for i in range(len(ch) - 1))
+        else:
+            cnots.extend((ph[i], ph[-1]) for i in range(len(ph) - 1))
+            cnots.extend((ch[i], ch[-1]) for i in range(len(ch) - 1))
+
+        while True:
+            # As per equations 9 and 10, if all the targets (resp. controls) are the same, the
+            # cnots commute. Further, if the control (resp. targets) are the same, the cnots can be
```

Comment:
I think the code is difficult to follow is partial because the data structure of CNOTs is hard to look up.
You can transfer that into two lists by `control_qubits, target_qubits = zip(*cnots)`. Now you can use built-in like`any`, `all` instead of nested for-loop and easier to read I think.

Reply:
Actually, I would need to flip control and target to get both equalities. Would renaming the loop that generate (x, y) instead clearer?

Reply:
Yes, I notice you flipped the control and target. But after I read it again, maybe we don't even need to worry about the flip at least for the first several lines (208-219). I think it can be simplified into this:
```
# find consecutive same elements
for i, cnot in enumerate(cnots):
    if i > 0 and cnot == cnots[i-1]:
        cnots.pop(i)
        cnots.pop(i-1)
        break
```
Do I understand this correctly?

Reply:
Yes, you understood it correctly, I think. However, re-reading the code, I think the simplification was not complete. If the CNOTs commute, then I should look at all the commuting CNOTs for identical targets/controls. I modified the code to reflect this. Sorry about this mistake.